packages/react-dom/src/server/ReactPartialRenderer.js

```diff
@@ -667,46 +669,65 @@ class ReactDOMServerRenderer {
     this.makeStaticMarkup = makeStaticMarkup;
 
     // Context (new API)
-    this.providerStack = []; // Stack of provider objects
-    this.providerIndex = -1;
+    this.contextIndex = -1;
+    this.contextStack = [];
+    this.contextValueStack = [];
+    if (__DEV__) {
+      this.contextProviderStack = [];
```

Comment:
Why'd we change the previous `providerStack` to `contextStack`?

Or maybe, why do we also need `contextProviderStack` in DEV? Couldn't our DEV warning just use:
```js
index > -1 &&
provider.type._context === (this.contextStack: any)[index],
```

Reply:
>Why'd we change the previous providerStack to contextStack?

From the PR:

>I also have a DEV-only stack for providers. I could've used it for context objects too, but I decided it's nice to avoid some property access on every pop when we need to read the context object. 

I meant that if I keep providers on the stack, I need to do a bunch of object access (`provider.type._context`) just to get to the context object. Storing it directly avoids that. I don't use provider itself anyway for anything other than DEV validation. This is similar to how we store Fibers on the stack only in DEV for validation.

>Or maybe, why do we also need contextProviderStack in DEV? Couldn't our DEV warning just use:
>index > -1 &&
>provider.type._context === (this.contextStack: any)[index],

This only validates that the context _type_ matches so it's less restrictive. We could have a bug where we accidentally pop the wrong provider object but miss it because it has the same type.

Reply:
> I also have a DEV-only stack for providers. I could've used it for context objects too, but I decided it's nice to avoid some property access on every pop when we need to read the context object.

It's not obvious to me that the cost of maintaining a second stack is less than the cost of a property access. I guess it's not a big deal either way, since it's DEV only though.

> This only validates that the context type matches so it's less restrictive. We could have a bug where we accidentally pop the wrong provider object but miss it because it has the same type.

Context (type), Provider, and Consumer have a 1-to-1-to-1 relationship, no?

Reply:
>It's not obvious to me that the cost of maintaining a second stack is less than the cost of a property access. I guess it's not a big deal either way, since it's DEV only though.

Exactly, I was mostly microoptimizing for prod.

>Context (type), Provider, and Consumer have a 1-to-1-to-1 relationship, no?

"Provider" is an object in this case. I guess `providerElement` would have been a clearer naming. So it's literally `<MyProvider />` that gets pushed and popped, not `MyProvider`.

Reply:
Ah! I see what you're saying now (for both cases) 😄 Makes sense. Thanks for explaining.